### PR TITLE
core: allow setting and propagation of eval priority on job de/registration 

### LINKED ISSUE #11434: Allow overriding evaluation priority for job commands and APIs
### Proposal

Jobs are dequeued by scheduler workers in [`priority`](https://www.nomadproject.io/docs/job-specification/job#priority) order. Priority also affects preemption and may affect other Nomad behaviors in the future (it is intentionally documented broadly so that it can be used whereever job priorities need to be weighed ... for better and worse).

As of Nomad 1.1.6 Job Registration evaluations are given the priority of the Job, but Deregistrations (`nomad job stop`) uses the default priority of 50.

#11426 changes the latter to also use the priority of the job for deregistrations to match user expectations.

As [Luiz proposed in a comment](https://github.com/hashicorp/nomad/pull/11426#pullrequestreview-795378043) ideally an operator would be able to override the **evaluation's** priority when submitting or stopping a job.

This issue proposes allowing setting the [`Evalutation.Priority`](https://pkg.go.dev/github.com/hashicorp/nomad/nomad/structs#Evaluation) field directly from the CLI and API:

```
nomad job run  -eval-priority=40 abc.nomad
nomad job stop -eval-priority=95 abc
```

### Use-cases

Nomad can have a large backlog of evaluations in busy clusters or when incidents occur. Operators have near-0 control over how that backlog is processed which can be tremendously frustrating when a high priority job seems "stuck" behind lower priority work because an operator attempted to stop it.

Since Nomad uses the `Job.Priority` field for preemption, it's also currently impossible to change scheduling priority without affecting preemption. Especially during incidents this is undesirable: an operator may have no idea how setting `Job.Priority = 100` affects preemption and just wants to get the job evaluated as quickly as possible.

### Attempted Solutions

There are no workarounds for controlling evaluation priority or much at all about evaluations.

### PR DESCRIPTION
**Pull request by jrasell**
This change allows setting the evaluation priority on job registration, update, and deregistration. In the event an override parameter is set, it will be validate to be within the 1-100 (inclusive) range. All evaluations triggered as a direct result of the job change request will inherit the override evaluation priority. This has meant the evaluation priority needs to be passed to the allocation reconciler and deployment watcher, as the eval priority and job priority are now distinctly different properties. If the operator does not set an override, the job priority is used which is consistent with the current behaviour.

I've added a [demo to my dev-mess repo](https://github.com/jrasell/dev-mess/tree/main/nomad/development/gh11434/job) which details one way to test this change and see the behaviour.

Each commit contains additional commentary where needed.

closes #11434

### CODE REVIEW

File: api/jobs.go [hunk 1]
```diff
@@ -290,14 +292,30 @@ type DeregisterOptions struct {
...
-		url.PathEscape(jobID), opts.Purge, opts.Global), &resp, q)
+
+	// The base endpoint to add query params to.
+	endpoint := "/v1/job/" + url.PathEscape(jobID)
+
+	// Protect against nil opts.
+	if opts != nil {
+		endpoint += fmt.Sprintf("?purge=%t&global=%t&eval_priority=%v",
+			opts.Purge, opts.Global, opts.EvalPriority)
+	}
```
Comment (tgross) [inline api/jobs.go:316]: Note to anyone else who reviews this, I went down a rabbithole looking if there was a nicer way of writing querystring params but [`url.Values`](https://pkg.go.dev/net/url#Values) expects string values so you end up `fmt.Sprintf`'ing in there anyways. 🤷

File: e2e/eval_priority/eval_priority.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	}
+	tc.jobIDs = []string{}
+
+	_, err := e2eutil.Command("nomad", "system", "gc")
+	f.NoError(err)
+}
+
+// TestJobRegisterWithoutEvalPriority makes sure that when not specifying an
+// eval priority on job register, the job priority is used.
+func (tc *EvalPriorityTest) TestJobRegisterWithoutEvalPriority(f *framework.F) {
```
Comment (tgross) [inline e2e/eval_priority/eval_priority.go:45]: I love that we're being thorough with E2E, but the priority set on the `Evaluation` object seems like it's already well-covered in the other tests you've written. If we're going to go through the "cost" of an E2E test, is it possible to exercise that the priorities are respected by the scheduler and deployment as we'd expect?

File: nomad/structs/structs.go [hunk 3]
```diff
@@ -8847,12 +8862,18 @@ type Deployment struct {
 	// status.
 	StatusDescription string
 
+	// EvalPriority tracks the priority of the evaluation which lead to the
+	// creation of this Deployment object. Any additional evaluations created
+	// as a result of this deployment can therefore inherit this value, which
+	// is not guaranteed to be that of the job priority parameter.
```
Comment (tgross) [inline nomad/structs/structs.go:8868]: Nice docstring describing the expectations 👍

File: nomad/deploymentwatcher/deployment_watcher.go [hunk 4]
```diff
@@ -827,7 +827,7 @@ func (w *deploymentWatcher) getEval() *structs.Evaluation {
 	return &structs.Evaluation{
 		ID:           uuid.Generate(),
 		Namespace:    w.j.Namespace,
-		Priority:     w.j.Priority,
+		Priority:     w.d.EvalPriority,
```
Comment (tgross) [inline nomad/deploymentwatcher/deployment_watcher.go:830]: During a server upgrade it's possible we end up with deployments created on the previous version that are then "watched" on a leader that's on the new version. So I think when we call `getEval` on that new leader we could get a deployment with an eval priority of 0. Maybe we should have some belt-and-suspenders here where we pick up the job priority if the `EvalPriority` is 0?

**Review by tgross — COMMENTED** [review]
Overall this looks great @jrasell! The way you broke up the PR into commits with good messages really helped expedite review here too. Thanks!

I've left a few comments/questions but I think the only thing that's potentially a blocker is my concern about deployments during an upgrade.

File: nomad/job_endpoint.go [hunk 5]
```diff
@@ -829,22 +837,23 @@ func (j *Job) Deregister(args *structs.JobDeregisterRequest, reply *structs.JobD
...
-	// the default value. Jobs with a lower than default register priority
-	// would therefore have their deregister eval priorities higher than
-	// expected.
-	priority := structs.JobDefaultPriority
-	if job != nil {
-		priority = job.Priority
-	}
-
 	// If the job is periodic or parameterized, we don't create an eval.
 	if job == nil || !(job.IsPeriodic() || job.IsParameterized()) {
```
[the remark this answers is not in the document]
Comment (jrasell) [reply nomad/job_endpoint.go:841]: Yeah I had this in the back of my mind but wasn't sure if we wanted to allow setting the eval priority on everything or just the identified critical job update actions. As you mention though, we can decide and follow-up with another PR if needed.

File: api/jobs.go (hunk 1 above)
Comment (jrasell) [reply api/jobs.go:316]: I'll add a comment on this line; I too spent time on trying to write this nicer.

File: e2e/eval_priority/eval_priority.go (hunk 2 above)
Comment (jrasell) [reply e2e/eval_priority/eval_priority.go:45]: Yeah I certainly got tunnel vision here. I'll rework the E2E as per your suggestion.

**Review by tgross — APPROVED** [review]
LGTM!

**Pull request merged**